### PR TITLE
Implement vread for ReadFileInputStream

### PR DESCRIPTION
**Pull request by zhli1142015**
Implement vread for ReadFileInputStream, #4833

### CODE REVIEW

**Comment by Yuhta** [comment]
We have recently added `vread` to `ReadFile`, please use or override the method there.  `InputStream` is only meant to be a thin wrapper of `ReadFile` going forward.

**Comment by zhli1142015** [comment]
> We have recently added `vread` to `ReadFile`, please use or override the method there. `InputStream` is only meant to be a thin wrapper of `ReadFile` going forward.

Hello @Yuhta ,
could you please point me the `vread` method you mentioned? I actually can't find it. I understand `InputStream` should be a thin wrapper, this change just let parallel loading from `BufferedInput.load` to `ReadFile.preadv`working as expected.
Thanks.

**Comment by Yuhta** [comment]
Sorry it is `preadv`, but functionally equivalent here:
https://github.com/facebookincubator/velox/blob/1678f5fd98a247fe8ee243da2925df1832e960b5/velox/common/file/File.h#L83

**Comment by zhli1142015** [comment]
> Sorry it is `preadv`, but functionally equivalent here:
> 
> https://github.com/facebookincubator/velox/blob/1678f5fd98a247fe8ee243da2925df1832e960b5/velox/common/file/File.h#L83

Thanks, @Yuhta , should i use `preadv(const std::vector<ReadFile::Segment>& segments)` in my pr, instead of `preadv(uint64_t offset, const std::vector<folly::Range<char*>>& buffers)`, is this correct?

**Comment by Yuhta** [comment]
You can use the new `preadv` to reimplement `InputStream::vread`.  The coalescing part you should handle in a subclass of `ReadFile` in a separate PR.  Which underlying `ReadFile` implementation you are trying to do the coalescing?

**Comment by zhli1142015** [comment]
> You can use the new `preadv` to reimplement `InputStream::vread`. The coalescing part you should handle in a subclass of `ReadFile` in a separate PR. Which underlying `ReadFile` implementation you are trying to do the coalescing?

We use ABFS (azure) file system now it's maintained in our repo. 
I think the new `preadv` is more suitable for our case, we don't need to fill gaps with nullptr in `InputStream` and remove them when reading data in file system.

**Comment by zhli1142015** [comment]
Hello @Yuhta ,
Could you please help review this PR?
Thanks.

File: velox/dwio/common/InputStream.cpp [hunk 1]
```diff
@@ -127,6 +127,33 @@ bool ReadFileInputStream::hasReadAsync() const {
   return readFile_->hasPreadvAsync();
 }
 
+void ReadFileInputStream::vread(
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/dwio/common/InputStream.cpp:130]: Hello @Yuhta ,
Thanks for you suggestion, I prefer to override the function in `ReadFileInputStream` for calling the new method `readFile_->preadv(segments)` you mentioned. Please let me know if I misunderstand your point, thanks.

[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/dwio/common/InputStream.cpp:130]: Bump this thread~
Hi @Yuhta , Please le me know what's the next step for the two `preadv` methods. If we want to migrate all callers to the new one in future, i'm ok to change the base class to consume `read(buffers)`. If these two functions would exist for long time, I would still prefer to implement it in `ReadFileInputStream` which can access `velox::ReadFile` and call new `preadv` directly, as `readv`'s signature is more alignable with new `preadv` method's.

[the remark this answers is not in the document]
Comment (Yuhta) [reply velox/dwio/common/InputStream.cpp:130]: Sorry for the confusion, I mean you can change `InputStream::vread` directly

File: velox/dwio/common/InputStream.cpp [hunk 2]
```diff
@@ -127,29 +127,39 @@ bool ReadFileInputStream::hasReadAsync() const {
...
     const std::vector<Region>& regions,
     const LogType purpose) {
   const auto size = buffers.size();
-  // the default implementation of this is to do the read sequentially
   DWIO_ENSURE_GT(size, 0, "invalid vread parameters");
   DWIO_ENSURE_EQ(regions.size(), size, "mismatched region->buffer");
-
-  // convert buffer to IOBufs and convert regions to VReadIntervals
-  LOG(INFO) << "[VREAD] fall back vread to sequential reads.";
+  std::vector<ReadFile::Segment> segments;
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/dwio/common/InputStream.cpp:137]: Thanks, added.

**Review by DanielMunozT — APPROVED** [review]

**Comment by zhli1142015** [comment]
Thanks @DanielMunozT for approving this PR, could this PR be merged?

**Pull request closed without merging**